app/gui/view/graph-editor/src/component/node/input/area.rs

```diff
@@ -884,26 +649,275 @@ impl Area {
+
+            // === View Mode ===
+
+            frp.output.source.view_mode <+ frp.set_view_mode;
+
+            in_profiling_mode <- frp.view_mode.map(|m| m.is_profiling());
+            finished          <- frp.set_profiling_status.map(|s| s.is_finished());
+            profiled          <- in_profiling_mode && finished;
+
+            use theme::code::syntax;
+            let std_selection_color      = model.styles_frp.get_color(syntax::selection);
+            let profiled_selection_color = model.styles_frp.get_color(syntax::profiling::selection);
+            let std_base_color           = model.styles_frp.get_color(syntax::base);
+            let profiled_base_color      = model.styles_frp.get_color(syntax::profiling::base);
+
+            selection_color_rgba <- profiled.switch(&std_selection_color,&profiled_selection_color);
+
+            selection_color.target          <+ selection_color_rgba.map(|c| color::Lcha::from(c));
+            model.label.set_selection_color <+ selection_color.value.map(|c| color::Lch::from(c));
+
+            init_colors         <- source::<()>();
+            std_base_color      <- all(std_base_color,init_colors)._0();
+            profiled_base_color <- all(profiled_base_color,init_colors)._0();
+            base_color          <- profiled.switch(&std_base_color,&profiled_base_color);
+            eval base_color ((color) model.label.set_property_default(color));
+            init_colors.emit(());
+        }
+
+        Self { frp, model }
+    }
+
+    #[allow(missing_docs)] // FIXME[everyone] All pub functions should have docs.
+    pub fn port_offset(&self, crumbs: &[Crumb]) -> Option<Vector2<f32>> {
+        let expr = self.model.expression.borrow();
+        expr.root_ref().get_descendant(crumbs).ok().map(|node| {
+            let unit = GLYPH_WIDTH;
+            let range_before = enso_text::Range::new(ByteDiff(0), node.payload.index);
+            let char_offset = expr.viz_code[range_before].chars().count();
+            let char_count = expr.viz_code[node.payload.range()].chars().count();
+            let width = unit * (char_count as f32);
+            let x = width / 2.0 + unit * (char_offset as f32);
+            Vector2::new(TEXT_OFFSET + x, 0.0)
+        })
+    }
+
+    #[allow(missing_docs)] // FIXME[everyone] All pub functions should have docs.
+    pub fn port_type(&self, crumbs: &Crumbs) -> Option<Type> {
+        let expression = self.model.expression.borrow();
+        expression.span_tree.root_ref().get_descendant(crumbs).ok().and_then(|t| t.tp.value())
+    }
+
+    #[allow(missing_docs)] // FIXME[everyone] All pub functions should have docs.
+    pub fn get_crumbs_by_id(&self, id: ast::Id) -> Option<Crumbs> {
+        self.model.id_crumbs_map.borrow().get(&id).cloned()
+    }
+
+    #[allow(missing_docs)] // FIXME[everyone] All pub functions should have docs.
+    pub fn label(&self) -> &text::Text {
+        &self.model.label
+    }
+
+    /// Set a scene layer for text rendering.
```

Comment:
if you are changing this code, please fix fixmes.

Reply:
All fixes are removed or supplied with a link to the pivotal issue.